Rework and Generalize History Import and Export

jmchilton opened the pull request:
The tracking PR for remote execution work (https://github.com/galaxyproject/galaxy/pull/7058) bumped into the problem of needing to serialize/deserialize Galaxy models to/from workers. Pickling was okay initially but doesn't solve a lot of the merging things back problem. In this standing branch I'm building up a library for dealing with model object serialization and deserialization effectively. Rather than write this from scratch - I'm re-worked history import/export to be a natural first use case of this framework.

Currently this branch includes the following:

- Allow import and export collections.
- Fix job I/O for separated tool and job script streams (merged in #7095 during 19.05)
- Fix job state handling so that all imported jobs have terminal state.
- Imported job parameters were not usable I think (based on workflow extraction testing), they now work and no longer require the toolbox on import or export (so if tools get added later after the import or removed before the export - I believe everything should still work).
- Fixes to eliminate HID handling for associations (objects were previously tracked by HID but inputs/outputs in history may not be in history - so HID is not at all appropriate, problem was amplified by collections).
- Fix job output mapping - was using the HDA name instead of job output name during import (old exports don't have this info - only new ones do).
- Fixes assigning HIDs to datasets during import and fixes to hid_counter - in many ways this was very broken for imported histories.
- Introduce cohesive model object import/export abstractions including a lot of new, well tested functionality.
  - Allow import/export of datasets and collections stand-alone regardless of history.
  - Allow export and re-import of datasets and collections (so for instance datasets can be created for a job and read in at the end of a job and merged back into the database). This is needed for extended metadata collection.
  - Separate implementation for import abstraction that allow import of legacy archives without complicating logic for current archive format.
  - Implementation of import and export abstraction for [BagIt](https://datatracker.ietf.org/doc/rfc8493/) using [bdbag](https://github.com/fair-research/bdbag).
  - Context manager support and clean APIs for all of this including import and export of directories, archives, and bags.
- A lot more integration testing - in particular most of our workflow extraction tests now export and reimport the history to ensure reproducibility survives history import/export.
- A lot more unit testing.

hexylena commented:
Some old history import/export tickets we can close after this :) https://github.com/galaxyproject/galaxy/issues/6120 https://github.com/galaxyproject/galaxy/issues/987

jmchilton committed:
Revamp history import/export with dozens of fixes and enhancements.

- Allow import and export collections.
- Fix job I/O for separated tool and job script streams (merged in #7095 during 19.05)
- Fix job state handling so that all imported jobs have terminal state.
- Fixes to eliminate HID handling for associations (objects were previously tracked by HID but inputs/outputs in history may not be in history - so HID is not at all appropriate, problem was amplified by collections).
- Fix job output mapping - was using the HDA name instead of job output name during import (old exports don't have this info - only new ones do).
- Fixes assigning HIDs to datasets during import and fixes to hid_counter - in many ways this was very broken for imported histories.
- Introduce cohesive model object import/export abstractions including a lot of new, well tested functionality.
  - Allow import/export of datasets and collections stand-alone regardless of history.
  - Allow export and re-import of datasets and collections (so for instance datasets can be created for a job and read in at the end of a job and merged back into the database). This is needed for extended metadata collection.
  - Separate implementation for import abstraction that allow import of legacy archives without complicating logic for current archive format.
  - Implementation of import and export abstraction for [BagIt](https://datatracker.ietf.org/doc/rfc8493/) using [bdbag](https://github.com/fair-research/bdbag).
  - Context manager support and clean APIs for all of this including import and export of directories, archives, and bags.
- A lot more unit testing.

FredericBGA commented:
> 
> 
> Some old history import/export tickets we can close after this :) #6120 #987

and this PR as well #4532

mvdbeek commented:
The unit test failures look relevant
`sqlalchemy.orm.exc.FlushError: New instance <HistoryDatasetCollectionAssociation at 0x7ff75f540550> with identity key (<class 'galaxy.model.HistoryDatasetCollectionAssociation'>, (1,), None) conflicts with persistent instance <HistoryDatasetCollectionAssociation at 0x7ff75f211630>`

jmchilton commented:
This was green before I rebased some more stuff into here this morning. I'll put this back in WIP until I make some progress on this.

jmchilton commented:
Back out of WIP.

mvdbeek commented:
@galaxybot test py3

mvdbeek commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -0,0 +1,1321 @@
+            self.app = app
+            self.security = app.security
+        else:
+            self.security = IdEncodingHelper(id_secret="randomdoesntmatter")
```
Is this for future work ? In the current state we don't test or use this, right ?

jmchilton replied:
Definitely used extensively in extended metadata collection #7058 but #7684 which is closer uses this (https://github.com/galaxyproject/galaxy/pull/7684/commits/65f747338a75ce205f6375ba4d7249884b29c8ba#diff-c96649080f6e1af45ac25198b92a4d68R28). So unit tests are coming quickly ideally.

mvdbeek commented on lib/galaxy/model/store/__init__.py:
```diff
@@ -0,0 +1,1321 @@
+                    shutil.copyfile(src, dest)
+
+        export_directory = self.export_directory
+        if dataset.id in self.included_datasets:
```
is there a case where this is not true ? Those would fail at `if file_name:` on line 998 I think

jmchilton replied:
So it would seem, good catch. I've eliminated the conditional in https://github.com/galaxyproject/galaxy/pull/7684 (which already had a bunch of model store fixes/enhancements).

mvdbeek reviewed (approved):
This is awesome, it's working well and is so much cleaner than what we had before, plus a great set of tests! I've added some minor stuff in https://github.com/jmchilton/galaxy/pull/72

mvdbeek commented:
Feel free to merge, I can also add https://github.com/jmchilton/galaxy/pull/72 as a separate PR

jmchilton commented:
@mvdbeek I've included jmchilton#72 as part of #7684 (which already had a bunch of model store fixes/enhancements), is that sufficient?

jmchilton commented:
Seems yes, thank you for the merge 😄. I'll write some tests for that tag handling stuff - it was on the TODO list anyway.

mvdbeek commented:
Yep, I saw you pulled it into https://github.com/galaxyproject/galaxy/pull/7684. Thanks again for spending time on this, I think this was much needed!
